File: ObjectTracking-fast/fast_tracking.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include "opencv2/features2d.hpp"
#include <opencv2/xfeatures2d.hpp>


// FAST ALGORITHM


using namespace cv;
using namespace std;

// random values for user-defined rectangle
int x_min = 36000;
int y_min = 36000;
int x_max = 0;
int y_max = 0;


// video path
std::string video_path="../../videos/bird.mp4";


// create FAST Detector
cv::Ptr<cv::FastFeatureDetector> fast = cv::FastFeatureDetector::create();
// create using BRIEF  description --> Compute descriptors
cv::Ptr<cv::xfeatures2d::BriefDescriptorExtractor> brief = cv::xfeatures2d::BriefDescriptorExtractor::create();


void CallBackFunc(int event, int x, int y, int flags, void* userdata)
{
    // Access the frame passed as userdata
    Mat* framePtr = static_cast<Mat*>(userdata);

    if (event == EVENT_LBUTTONDOWN)
    {
        cout << "Right button of the  is clicked - position (" << x << ", " << y << ")" << endl;

        // Update x_min, y_min, x_max, y_max based on mouse click position
        x_min = min(x, x_min);
        y_min = min(y, y_min);
        x_max = max(x, x_max);
        y_max = max(y, y_max);

        // Draw rectangle on the frame
        rectangle(*framePtr, Point(x_min, y_min), Point(x_max, y_max), Scalar(255, 255, 0), 5, LINE_8);

        // Display the updated frame
        imshow("Video Player", *framePtr);
    }
}


void display_points(const Mat& descriptors, const vector<KeyPoint>& keypoints)
{

    // Open video file
    VideoCapture video(video_path);


    // Loop through video frames
    Mat frame;

    while (video.read(frame)) {


        Mat gray_frame;

        resize(frame, frame, Size(), 0.4, 0.4, INTER_CUBIC);

        cvtColor(frame, gray_frame, COLOR_BGR2GRAY);

        // Detect keypoints
        vector<KeyPoint> keypoints2;
        Mat saved_descriptors2;

        fast->detect(frame, keypoints2); 
        brief->compute(frame, keypoints2, saved_descriptors2);

        // Match features using BFMatcher (Brute-Force Matcher)
        cv::BFMatcher matcher(cv::NORM_HAMMING, true);  // Use Hamming distance for binary descriptors like ORB
        std::vector<cv::DMatch> matches;
        matcher.match(descriptors, saved_descriptors2, matches);


        // Draw circles on matched points
        for (size_t i = 0; i < matches.size(); i++) {
            int idx1 = matches[i].queryIdx;  // Index of matched keypoint in image1
            int idx2 = matches[i].trainIdx;  // Index of matched keypoint in image2

            // Get the matched keypoints
            cv::Point2f pt1 = keypoints[idx1].pt;
            cv::Point2f pt2 = keypoints2[idx2].pt;

            // Draw a circle at each matching keypoint
            cv::circle(frame, pt2, 5, cv::Scalar(0, 255, 0), 2);  // Green circle on img1
        }


        // Display frame
        imshow("Video", frame);

        // Break if 'Esc' key is pressed
        if (waitKey(30) == 27) {
            break;
        }
    }

    // Release video capture object
    video.release();
    destroyAllWindows();

}


void take_roi(void* frame)
{

    Mat* framePtr = static_cast<Mat*>(frame);

    // Create the rectangle (x, y, width, height)
    Rect roi(x_min, y_min, x_max - x_min, y_max - y_min);

    // Extract the ROI
    Mat image_roi = (*framePtr)(roi);

    // Do something with the extracted ROI, for example, display it
    imshow("ROI", image_roi);
    waitKey(0); // Wait indefinitely until a key is pressed

    // Convert ROI to grayscale
    Mat gray_roi;
    cvtColor(image_roi, gray_roi, COLOR_BGR2GRAY);


    // Detect keypoints
    vector<KeyPoint> keypoints;
    Mat saved_descriptors;

    fast->detect(image_roi, keypoints);  
    brief->compute(image_roi, keypoints, saved_descriptors);

    display_points(saved_descriptors,keypoints);

}



int main() {

    // create VideoCapture object
    VideoCapture video(video_path);

    // Check if the video file was opened successfully
    if (!video.isOpened()) {
        std::cout << "Error: Could not open the video file." << std::endl;
        return -1;
    }

    // Create a window to display the video
    namedWindow("Video Player", WINDOW_NORMAL);
    resizeWindow("Video Player", Size(video.get(CAP_PROP_FRAME_WIDTH), video.get(CAP_PROP_FRAME_HEIGHT)));

    // create Mat object for first frame
    Mat frame;

    // Read the first frame
    if (!video.read(frame)) {
        std::cout << "Error: Could not read the first frame." << std::endl;
        return -1;
    }

    // Set the callback function and pass frame as userdata
    setMouseCallback("Video Player", CallBackFunc, &frame);

    // Display the first frame
    imshow("Video Player", frame);

    // Wait for a key press
    waitKey(0);

    // Release the VideoCapture object and close the display window
    video.release();
    destroyAllWindows();

    take_roi(&frame);

    return 0;
}

File: ObjectTracking-fast/fast_tracking.py
import cv2
import numpy as np 
import matplotlib.pyplot as plt
import time



# Path to video  
video_path=r"../videos/plane.mp4"
video = cv2.VideoCapture(video_path)

# read only the first frame for drawing a rectangle for the desired object
ret,frame = video.read()

# STEP 1 : Chosing the target object with mouse by drawing rectangle around it.

# I am giving  big random numbers for x_min and y_min because if you initialize them as zeros whatever coordinate you go minimum will be zero 
x_min,y_min,x_max,y_max=36000,36000,0,0


def coordinat_chooser(event,x,y,flags,param):
    global go , x_min , y_min, x_max , y_max

    # when you click the right button, it will provide coordinates for variables
    if event==cv2.EVENT_LBUTTONDOWN:
        
        # if current coordinate of x lower than the x_min it will be new x_min , same rules apply for y_min 
        x_min=min(x,x_min) 
        y_min=min(y,y_min)

         # if current coordinate of x higher than the x_max it will be new x_max , same rules apply for y_max
        x_max=max(x,x_max)
        y_max=max(y,y_max)

        # draw rectangle
        cv2.rectangle(frame,(x_min,y_min),(x_max,y_max),(0,255,0),1)


    """
        if you didn't like your rectangle (maybe if you made some misscliks),  reset the coordinates with the middle button of your mouse
        if you press the middle button of your mouse coordinates will reset and you can give a new 2-point pair for your rectangle
    """
    if event==cv2.EVENT_MBUTTONDOWN:
        print("reset coordinate  data")
        x_min,y_min,x_max,y_max=36000,36000,0,0

cv2.namedWindow('coordinate_screen')
# Set mouse handler for the specified window, in this case, "coordinate_screen" window
cv2.setMouseCallback('coordinate_screen',coordinat_chooser)


while True:
    cv2.imshow("coordinate_screen",frame) # show only first frame 
    
    k = cv2.waitKey(5) & 0xFF # after drawing rectangle press ESC   
    if k == 27:
        cv2.destroyAllWindows()
        break


# STEP 2 :  Extract Features from the target object ( not from full image )

# take region of interest ( take inside of rectangle )
roi_image=frame[y_min:y_max,x_min:x_max]

# convert roi to grayscale, SIFT Algorithm works with grayscale images
roi_gray=cv2.cvtColor(roi_image,cv2.COLOR_BGR2GRAY) 

# Initialize the FAST detector and BRIEF descriptor extractor
fast = cv2.FastFeatureDetector_create(threshold=20)
brief = cv2.xfeatures2d.BriefDescriptorExtractor_create()

# detect keypoints
keypoints_1 = fast.detect(roi_gray, None)
# descriptors
keypoints_1, descriptors_1 = brief.compute(roi_gray, keypoints_1)

# draw keypoints for visualizing
keypoints_image = cv2.drawKeypoints(roi_image, keypoints_1, outImage=None, color=(0, 255, 0))
# display keypoints
plt.imshow(keypoints_image,cmap="gray")


# STEP 3 : Track Objects by using FAST Algorithm

# matcher object
bf = cv2.BFMatcher()

# Variables for FPS calculation
frame_count = 0
start_time = time.time()
     
while True :
    # reading video 
    ret,frame=video.read()

    if ret:
          # convert frame to gray scale 
        frame_gray=cv2.cvtColor(frame,cv2.COLOR_BGR2GRAY)
    
        
        # Detect keypoints using FAST
        keypoints_2 = fast.detect(frame_gray, None)

        # Compute descriptors using BRIEF
        keypoints_2, descriptors_2 = brief.compute(frame_gray, keypoints_2)
        
        """
        Compare the keypoints/descriptors extracted from the 
        first frame(from target object) with those extracted from the current frame.
        """
        if descriptors_2 is  not None:
            matches =bf.match(descriptors_1, descriptors_2)
    
    
            for match in matches:
            
                # queryIdx gives keypoint index from target image
                query_idx = match.queryIdx
                
                # .trainIdx gives keypoint index from current frame 
                train_idx = match.trainIdx
                
                # take coordinates that matches
                pt1 = keypoints_1[query_idx].pt
                
                # current frame keypoints coordinates
                pt2 = keypoints_2[train_idx].pt
                
                # draw circle to pt2 coordinates , because pt2 gives current frame coordinates
                cv2.circle(frame,(int(pt2[0]),int(pt2[1])),5,(255,0,0),-1)
    
        # Calculate and display FPS
        frame_count += 1
        elapsed_time = time.time() - start_time
        fps = frame_count / elapsed_time
        cv2.putText(frame, f"FPS: {fps:.2f}", (10, 30), cv2.FONT_HERSHEY_SIMPLEX, 1, (0, 255, 0), 3)
        
    
        cv2.imshow("coordinate_screen",frame) 
    
    
        k = cv2.waitKey(5) & 0xFF # after drawing rectangle press esc   
        if k == 27:
            cv2.destroyAllWindows()
            break
    else:
        break

video.release()
cv2.destroyAllWindows()